### PR TITLE
Measurements/Core/Feat: Add new stream to stream calculated into gold

### PR DESCRIPTION
**Pull request by Gizmooa**
# Description

Creates an entry point that starts streaming from an external view, and into the gold layer.

# References

Infra PR: https://github.com/Energinet-DataHub/dh3-infrastructure/pull/3216

### CODE REVIEW

File: source/core/tests/helpers/builders/calculated_builder.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    ):
+        self.data.append(
+            (
+                metering_point_id,
+                metering_point_type,
+                observation_time,
+                orchestration_instance_id,
+                orchestration_type,
+                quantity,
+                quantity_quality or random.choice(["measured", "estimated", "calculated", "missing"]),
```
Comment (ShaBloze) [inline source/core/tests/helpers/builders/calculated_builder.py:35]: move the `random_choice` to line 21 as default. 

If we want to add a test where the `quantity_quality` is None, then it is currently not possible.

File: source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+
+def test__stream_measurements_calculated_to_gold__calls_expected(spark: SparkSession):
+    # Arrange
+    calculated_repo_mock = Mock(spec=CalculatedMeasurementsRepository)
+    gold_stream_mock = Mock(spec=GoldMeasurementsStream)
+    with (
+        patch.object(sut, "CalculatedMeasurementsRepository", return_value=calculated_repo_mock),
+        patch.object(sut, "GoldMeasurementsStream", return_value=gold_stream_mock),
+    ):
```
Comment (ShaBloze) [inline source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py:18]:
```suggestion
def test__stream_measurements_calculated_to_gold__calls_expected(spark: SparkSession, mocker: MockerFixture):
    # Arrange
    calculated_repo_mock = Mock(spec=CalculatedMeasurementsRepository)
    gold_stream_mock = Mock(spec=GoldMeasurementsStream)
    mocker.patch.object(sut, "CalculatedMeasurementsRepository", return_value=calculated_repo_mock)
    mocker.patch.object(sut, "GoldMeasurementsStream", return_value=gold_stream_mock)
```

use the MockerFixture instead. Goes for all tests with mocking

File: source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        calculated_repo_mock.read_stream.assert_called_once()
+        gold_stream_mock.write_stream.assert_called_once_with(
+            "measurements_calculated",
+            "measurements_calculated_to_gold",
+            calculated_repo_mock.read_stream.return_value,
+            sut._batch_operation,
+        )
+
+
+def test__pipeline_measurements_calculated_to_gold__calls_append_to_gold_measurements(spark: SparkSession):
```
Comment (ShaBloze) [inline source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py:34]:
```suggestion
def test__batch_operation__calls_append_to_gold_measurements(spark: SparkSession):
```

File: source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py
Comment (ShaBloze) [inline source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py:1]: type in file name, rename to `test_calculated_measurements_stream.py`

File: source/core/tests/unit_tests/gold/infrastructure/streams/test_measurements_gold_stream.py
Comment (ShaBloze) [inline source/core/tests/unit_tests/gold/infrastructure/streams/test_measurements_gold_stream.py:1]: should we add a test for checkpoint and query name calculated_to_gold?

File: source/core/src/core/gold/infrastructure/config/external_database_names.py [hunk 4]
```diff
@@ -0,0 +1,2 @@
+class ExternalDatabaseNames:
+    calculated = "measurements_calculated"
```
[the remark this answers is not in the document]
Comment (Gizmooa) [reply source/core/src/core/gold/infrastructure/config/external_database_names.py:2]: This is a part of Wholesale, I discussed it with them in terms of them making an env variable. But they've done it with hardcoded strings so they're not creating it. So I ended up going this way for now. We could push more if we like.

File: source/core/tests/unit_tests/gold/infrastructure/streams/test_measurements_gold_stream.py
Comment (Gizmooa) [reply source/core/tests/unit_tests/gold/infrastructure/streams/test_measurements_gold_stream.py:1]: It has been done in source/core/tests/unit_tests/gold/infrastructure/test_calculated_measurements_repository.py,  at least with query name. Cant remember if we have checked checkpoint name before

File: source/core/src/core/gold/domain/schemas/calculated_measurements.py
[the remark this answers is not in the document]
Comment (Gizmooa) [reply source/core/src/core/gold/domain/schemas/calculated_measurements.py:1]: Hmm, I dont think wholesale has added this to the package? But i can check. :)

File: source/core/tests/helpers/builders/calculated_builder.py (hunk 1 above)
Comment (Gizmooa) [inline source/core/tests/helpers/builders/calculated_builder.py:35]:
```suggestion
        quantity_quality=random.choice(["measured", "estimated", "calculated", "missing"]),
        quantity_unit: str = "kWh",
        resolution: str = "PT1H",
        transaction_creation_datetime=datetime.now(),
        transaction_id="",
    ):
        self.data.append(
            (
                metering_point_id,
                metering_point_type,
                observation_time,
                orchestration_instance_id,
                orchestration_type,
                quantity,
                quantity_quality,
```

Comment (Gizmooa) [reply source/core/tests/helpers/builders/calculated_builder.py:35]: I'll adjust :)

File: source/core/src/core/gold/domain/schemas/calculated_measurements.py
[the remark this answers is not in the document]
Comment (Gizmooa) [reply source/core/src/core/gold/domain/schemas/calculated_measurements.py:1]: Idk if it would make sense to add it for them if not? Maybe?

File: source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py
Comment (Gizmooa) [reply source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py:1]: adjusted

File: source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py (hunk 3 above)
Comment (Gizmooa) [reply source/core/tests/unit_tests/gold/application/streams/test_calclated_measurements_stream.py]: Adjusted

File: source/core/src/core/gold/domain/schemas/calculated_measurements.py
[the remark this answers is not in the document]
Comment (ShaBloze) [reply source/core/src/core/gold/domain/schemas/calculated_measurements.py:1]: It seems to be added [measurements calculated](https://github.com/Energinet-DataHub/opengeh-python-packages/blob/main/source/geh_common/src/geh_common/data_products/measurements_calculated/calculated_measurements_v1.py)

File: source/core/src/core/gold/infrastructure/config/external_database_names.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ShaBloze) [reply source/core/src/core/gold/infrastructure/config/external_database_names.py:2]: It is [here](https://github.com/Energinet-DataHub/dh3-infrastructure/blob/ae28556f4acb88d49793e5d4787fe594434cedac/measurements/terraform/core/main/dbw-unity.tf#L168) so we would use that in environment variable

File: source/core/src/core/gold/domain/schemas/calculated_measurements.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    StructField,
+    StructType,
+    TimestampType,
+)
+
+from core.gold.domain.constants.column_names.calculated_measurements_column_names import (
+    CalculatedMeasurementsColumnNames,
+)
+
+calculated_measurements_schema = StructType(
```
Comment (BjarkeMeier) [inline source/core/src/core/gold/domain/schemas/calculated_measurements.py:13]: I think this has been added to geh_common.

File: source/core/src/core/gold/domain/transformations/calculated_measurements_transformations.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import pyspark.sql.functions as F
+from pyspark.sql import DataFrame
+
+from core.gold.domain.constants.column_names.calculated_measurements_column_names import (
+    CalculatedMeasurementsColumnNames,
+)
+from core.gold.domain.constants.column_names.gold_measurements_column_names import GoldMeasurementsColumnNames
```
Comment (BjarkeMeier) [inline source/core/src/core/gold/domain/transformations/calculated_measurements_transformations.py:7]: Just curious: Do you have a class for gold names because you use different names in different stages, or because you want to include only columns used in gold?

File: source/core/src/core/gold/infrastructure/config/external_database_names.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/core/src/core/gold/infrastructure/config/external_database_names.py:2]: @djorgensendk and team fusion strongly prefers runtime dependencies over deployment dependencies between subsystems.

[the remark this answers is not in the document]
Comment (Gizmooa) [reply source/core/src/core/gold/infrastructure/config/external_database_names.py:2]: I see, I'll adjust (y)

File: source/core/src/core/gold/domain/schemas/calculated_measurements.py (hunk 5 above)
Comment (Gizmooa) [reply source/core/src/core/gold/domain/schemas/calculated_measurements.py:13]: Morten just pointed it out, I'll adjust it :)

File: source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+class CalculatedMeasurementsRepository:
+    def __init__(self) -> None:
+        calculated_schema_name = ExternalDatabaseNames.calculated
+        calculated_view_name = ExternalViewNames.hourly_calcualted_measurements_v1
+        self.view = f"{calculated_schema_name}.{calculated_view_name}"
+        self.spark = spark_session.initialize_spark()
+
+    def read_stream(self) -> DataFrame:
+        spark = spark_session.initialize_spark()
+        return spark.readStream.format("delta").option("ignoreDeletes", "true").table(self.view)
```
Comment (BjarkeMeier) [inline source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py:17]:
```suggestion
        return self.spark.readStream.format("delta").option("ignoreDeletes", "true").table(self.view)
```

File: source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+from core.gold.infrastructure.config.external_database_names import ExternalDatabaseNames
+from core.gold.infrastructure.config.external_view_names import ExternalViewNames
+
+
+class CalculatedMeasurementsRepository:
+    def __init__(self) -> None:
+        calculated_schema_name = ExternalDatabaseNames.calculated
+        calculated_view_name = ExternalViewNames.hourly_calcualted_measurements_v1
+        self.view = f"{calculated_schema_name}.{calculated_view_name}"
+        self.spark = spark_session.initialize_spark()
```
Comment (BjarkeMeier) [inline source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py:13]: Do you initialize spark sessions everywhere instead of passing the session to classes/functions where needed?

File: source/core/src/core/gold/domain/transformations/calculated_measurements_transformations.py (hunk 6 above)
Comment (Gizmooa) [reply source/core/src/core/gold/domain/transformations/calculated_measurements_transformations.py:7]: We've one for every layer, just to ensure if something where to change, then it is easily changeable with the transformations and selects 😄 Having to alias everywhere is a bit annoying tho 😆

File: source/core/src/core/gold/domain/constants/streaming/checkpoint_names.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+from enum import Enum
+
+
+class CheckpointNames(Enum):
```
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/core/src/core/gold/domain/constants/streaming/checkpoint_names.py:4]: The same question applies to `QueryNames`. I may be missing something, but it looks like you're just offering the possibility of using the class in the wrong way.

File: source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py (hunk 8 above)
Comment (Gizmooa) [reply source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py:13]: the initialize does a getOrCreate, so it will fetch an instance if exists, so no, it is not created every time. 😄

File: source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py (hunk 7 above)
Comment (Gizmooa) [reply source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py:17]: I was a bit too quick I see 😆

File: source/core/src/core/gold/domain/constants/streaming/checkpoint_names.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Gizmooa) [reply source/core/src/core/gold/domain/constants/streaming/checkpoint_names.py:4]: 'you're just offering the possibility of using the class in the wrong way.'
How exactly? 😄 

But I was also considering simply having a file with the CONST names inside, instead of encapsulating inside a class. It was just because this is how we've done normally 😄

File: source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py (hunk 8 above)
Comment (Gizmooa) [reply source/core/src/core/gold/infrastructure/repositories/calculated_measurements_repository.py:13]: Very true, also had considered if it should be renamed to indicate that it is a get or create happening :)

File: source/core/src/core/gold/domain/constants/streaming/checkpoint_names.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Gizmooa) [reply source/core/src/core/gold/domain/constants/streaming/checkpoint_names.py:4]: We've talked about this over teams. 😄

File: source/core/tests/helpers/schema_helper.py [hunk 10]
```diff
@@ -17,3 +19,12 @@ def create_schemas(spark: SparkSession) -> None:
...
     spark.sql(f"CREATE DATABASE IF NOT EXISTS {core_internal_settings.core_internal_database_name}")
+
+
+def create_external_schemas(spark: SparkSession) -> None:
+    calculated_settings = CalculatedSettings()
+    spark.sql(f"CREATE DATABASE IF NOT EXISTS {calculated_settings.calculated_database_name}")
+
+
+def create_external_tables(spark: SparkSession) -> None:
+    spark.sql(f"CREATE DATABASE IF NOT EXISTS {ExternalViewNames.calculated_measurements_v1}")
```
Comment (ShaBloze) [inline source/core/tests/helpers/schema_helper.py:30]:
```suggestion
    spark.sql(f"CREATE TABLE IF NOT EXISTS {ExternalViewNames.calculated_measurements_v1}")
```

**Review by ShaBloze — DISMISSED** [review]
LGTM

**Review by ShaBloze — APPROVED** [review]
